scripts/template_matcher.py

```diff
@@ -62,12 +63,73 @@ def _compute_prediction(self, k, img, kp, des):
         # TODO: find corresponding points in the input image and the templae image
         #       put keypoints from template image in template_pts
         #       put corresponding keypoints from input image in img_pts
-
-
-        #TODO: change img to img_T once you do the homography transform
-        visual_diff = compare_images(img, self.signs[k])
-        return visual_diff
+        img_kp, img_des = self.orb.detectAndCompute(img,None)
+
+        template_pts = np.array([[0.,0.]])
+        img_pts = np.array([[0.,0.]])
+        
+        for i in xrange(len(des[k])):
+            descriptor = des[k][i]
+            closest = [False, False]
+            second_closest = [False, False]
+            for j in xrange(len(img_des)):
+                img_descriptor = img_des[j]
+                dist = np.linalg.norm(descriptor - img_descriptor, 2)
+                if closest[0]:
+                    if dist < closest[1]:
+                        second_closest[0] = closest[0]
+                        second_closest[1] = closest[1]
+                        closest[0] = j
+                        closest[1] = dist
+                    elif (not second_closest[0]) or (dist < second_closest[1]):
+                        second_closest[0] = j
+                        second_closest[1] = dist
+                else:
+                    closest[0] = j
+                    closest[1] = dist
+            if closest[1] < (0.7*second_closest[1]):
+                template_pts = np.append(template_pts, [[kp[k][i].pt[0], kp[k][i].pt[1]]], axis=0)
+                img_pts = np.append(img_pts, [[img_kp[i].pt[0], img_kp[i].pt[1]]], axis=0)
+
+        if len(img_pts) > 2:
+            M, mask = cv2.findHomography(img_pts, template_pts, cv2.RANSAC, self.ransac_thresh)
+            img_T = cv2.warpPerspective(img, M, self.signs[k].shape[::-1])
+            visual_diff = compare_images(img_T, self.signs[k])
+            return visual_diff
+        else:
+            return compare_images(img, self.signs[k])
 # end of TemplateMatcher class
 
 def compare_images(img1, img2):
-    return 0
\ No newline at end of file
+    img1_mean = np.sum(img1)/float(len(img1)*len(img1[0]))
```

Comment:
You can also do this to get a single value for the mean intensity of your image.

img1_mean = img1.mean()